### PR TITLE
Tags List Refresh: Mark III

### PR DESCRIPTION
**Pull request by jleandroperez**
### Details:
In this PR we're enhancing the Tags List:

- New **Untagged Notes** filter
- The Tags section will no longer be displayed, whenever there are no tags
- Encapsulated all of the filtering NSPredicate(s)
- And we're getting new Unit Tests!

@danielebogo Daniele!! May I bug you with this one?
**Thank you!!!**

Ref. #408

### Testing: Hiding Tags Sections
1. Log into a testing account with no Tag

- [x] Verify that an account with no Tags doesn't show the `Tags` section
- [x] Verify that an account with no Tags doesn't show the `Untagged` section

### Testing: Predicates!
1. Log into your account
2. Add three notes (with random contents)
3. Tag one of the new notes
4. Trash one of the new notes
5. And of course, leave one of them untagged

- [x] Verify the `Trash` filter shows the trashed note
- [x] Verify the `Untagged` filter shows the untagged note
- [x] Verify that tapping over the tag you've set in (3) properly displays the new note
- [x] Verify the new unit tests look green!

---

### CODE REVIEW

File: SimplenoteTests/NSPredicateSimplenoteTests.swift [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        entity.tags = "[ \"second\", \"third\", \"" + tag + "\" ]"
+        XCTAssertTrue(NSPredicate.predicateForTag(with: tag).evaluate(with: entity))
+    }
+
+    /// Verifies that `NSPredicate.predicateForTag` properly deals with slashes
+    ///
+    func testPredicateForTagProperlyHandlesTagsWithSlashes() {
+        let tag = "\\Yosemite"
+        let entity = MockupEntity()
+        entity.tags = "[ \"\\\\Yosemite\" ]"
```
Comment (dmsnell) [inline SimplenoteTests/NSPredicateSimplenoteTests.swift]: yo yo yo

Comment (jleandroperez) [reply SimplenoteTests/NSPredicateSimplenoteTests.swift]: YOSEMITE!

Comment (dmsnell) [reply SimplenoteTests/NSPredicateSimplenoteTests.swift]: 🎶

File: Simplenote/Classes/NSPredicate+Simplenote.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ///
+    @objc(predicateForSearchText:)
+    static func predicateForSearchText(searchText: String) -> NSPredicate {
+        let words = searchText.trimmingCharacters(in: .whitespaces).components(separatedBy: .whitespaces)
+        var output = [NSPredicate]()
+
+        for word in words where !word.isEmpty {
+            let predicate = NSPredicate(format: "content CONTAINS[c] %@", word)
+            output.append(predicate)
+        }
```
Comment (danielebogo) [inline Simplenote/Classes/NSPredicate+Simplenote.swift]: Not sure if this is the case but if you trim the text for `.whitespacesAndNewlines` then you can avoid the `where !word.isEmpty`. So you can reduce the code with `let output = words.map { NSPredicate(format: "content CONTAINS[c] %@", $0) }`. Have you tried this already?

File: Simplenote/Classes/SPNoteListViewController+Extensions.swift [hunk 3]
```diff
@@ -72,8 +72,10 @@ extension SPNoteListViewController {
         let selectedTag = SPAppDelegate.shared().selectedTag ?? NSLocalizedString("All Notes", comment: "Title: No filters applied")
 
         switch selectedTag {
-        case kSimplenoteTagTrashKey:
+        case kSimplenoteTrashKey:
             title = NSLocalizedString("Trash-noun", comment: "Title: Trash Tag is selected")
+        case kSimplenoteUntaggedKey:
```
Comment (danielebogo) [inline Simplenote/Classes/SPNoteListViewController+Extensions.swift]: Do you really really really need the Hungarian notation? 😁

File: Simplenote/Classes/SPNoteListViewController.m [hunk 4]
```diff
@@ -722,73 +722,79 @@ - (void)update {
 
 - (void)updateFetchPredicate
 {
-    SPAppDelegate *appDelegate = [SPAppDelegate sharedDelegate];
-    if (appDelegate.selectedTag != nil && [appDelegate.selectedTag isEqualToString:kSimplenoteTagTrashKey]) {
-        
+    NSString *selectedTag = [[SPAppDelegate sharedDelegate] selectedTag];
+    if ([selectedTag isEqualToString:kSimplenoteTrashKey]) {
```
Comment (danielebogo) [inline Simplenote/Classes/SPNoteListViewController.m]: ☝️here as well

File: Simplenote/Classes/SPNoteListViewController.m [hunk 5]
```diff
@@ -722,73 +722,79 @@ - (void)update {
...
         tagFilterType = SPTagFilterTypeDeleted;
-    }
-    else {
-        
+    } else if ([selectedTag isEqualToString:kSimplenoteUntaggedKey]) {
```
Comment (danielebogo) [inline Simplenote/Classes/SPNoteListViewController.m]: ☝️here as well

File: Simplenote/Classes/SPNoteListViewController.m [hunk 6]
```diff
@@ -722,73 +722,79 @@ - (void)update {
...
+                break;
+            }
+
+            NSPredicate *predicate = [NSPredicate predicateForTagWith:appDelegate.selectedTag];
+            [predicateList addObject:predicate];
+            break;
+        }
+        default: {
+            break;
+        }
```
Comment (danielebogo) [inline Simplenote/Classes/SPNoteListViewController.m]: I think you can remove the `{ ... }` in most of the cases right?

File: Simplenote/Classes/SPTagsListViewController.m [hunk 7]
```diff
@@ -213,7 +220,7 @@ - (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)
 }
 
 - (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
-    return SPTagsListSectionCount;
+    return (self.numberOfTags == 0) ? kSPTagListEmptyStateSectionCount : SPTagsListSectionCount;
```
Comment (danielebogo) [inline Simplenote/Classes/SPTagsListViewController.m]: ☝️Hungarian notation here as well

File: Simplenote/Classes/SPTagsListViewController.m [hunk 8]
```diff
@@ -317,6 +331,10 @@ - (void)configureCell:(SPTagListViewCell *)cell atIndexPath:(NSIndexPath *)index
             [self configureTagCell:cell atIndexPath:indexPath];
             break;
         }
+        case SPTagsListSectionBottom: {
+            [self configureBottomCell:cell atIndexPath:indexPath];
+            break;
+        }
```
Comment (danielebogo) [inline Simplenote/Classes/SPTagsListViewController.m]: I think you can remove the `{ ... }` in this switch case too

File: Simplenote/Classes/SPTagsListViewController.m [hunk 9]
```diff
@@ -350,6 +368,13 @@ - (void)configureTagCell:(SPTagListViewCell *)cell atIndexPath:(NSIndexPath *)in
     cell.delegate = self;
 }
 
+- (void)configureBottomCell:(SPTagListViewCell *)cell atIndexPath:(NSIndexPath *)indexPath {
+    NSString *text = NSLocalizedString(@"Untagged Notes", @"Allows selecting notes with no tags");
+
+    cell.tagNameTextField.text = text;
```
Comment (danielebogo) [inline Simplenote/Classes/SPTagsListViewController.m]: You can assign the _localized string_ directly in this case

File: Simplenote/Classes/SPTagsListViewController.m [hunk 10]
```diff
@@ -371,6 +396,9 @@ - (BOOL)shouldSelectCellAtIndexPath:(NSIndexPath *)indexPath {
         case SPTagsListSectionTags: {
             return selectedTag == [self tagAtTableViewIndexPath:indexPath].name;
         }
+        case SPTagsListSectionBottom: {
+            return selectedTag == kSimplenoteUntaggedKey;
+        }
```
Comment (danielebogo) [inline Simplenote/Classes/SPTagsListViewController.m]: I think you can avoid the `{ ... }` in the _SPTagsListSectionTags_ and _SPTagsListSectionBottom_, and return directly the value. Also in nested switch as well.

Comment (danielebogo) [reply Simplenote/Classes/SPTagsListViewController.m]: Oh and the Hungarian Notation too 😊

File: Simplenote/Classes/SPTagsListViewController.m [hunk 11]
```diff
@@ -411,13 +439,18 @@ - (void)didSelectTagAtIndexPath:(NSIndexPath *)indexPath {
     }
 }
 
+- (void)didSelectBottomRowAtIndex:(NSIndexPath *)indexPath {
+    [SPTracker trackListUntaggedViewed];
+    [self openNoteListForTagName:kSimplenoteUntaggedKey];
```
Comment (danielebogo) [inline Simplenote/Classes/SPTagsListViewController.m]: Hungarian notation here as well ☝️

File: Simplenote/Classes/SPTagsListViewController.m [hunk 12]
```diff
@@ -411,13 +439,18 @@ - (void)didSelectTagAtIndexPath:(NSIndexPath *)indexPath {
...
+}
+
 - (void)allNotesWasPressed {
     [self openNoteListForTagName:nil];
 }
 
 - (void)trashWasPressed {
     [SPTracker trackTrashViewed];
-    [self openNoteListForTagName:kSimplenoteTagTrashKey];
+    [self openNoteListForTagName:kSimplenoteTrashKey];
```
Comment (danielebogo) [inline Simplenote/Classes/SPTagsListViewController.m]: Hungarian notation here as well ☝️

File: Simplenote/SPConstants.m [hunk 13]
```diff
@@ -42,7 +42,11 @@
 NSString *const kFirstLaunchKey                     = @"SPFirstLaunch";
 NSString *const kSelectedTagKey                     = @"SPSelectedTag";
 NSString *const kSelectedNoteKey                    = @"SPSelectedNote";
-NSString *const kSimplenoteTagTrashKey              = @"trash";
+NSString *const kSimplenoteTrashKey                 = @"__##__trash__##__";
+NSString *const kSimplenoteUntaggedKey              = @"__##__untagged__##__";
+NSString *const kSimplenoteSystemTagPinned          = @"pinned";
+NSString *const kSimplenoteSystemTagShared          = @"shared";
+NSString *const kSimplenoteSystemTagUnread          = @"unread";
```
Comment (danielebogo) [inline Simplenote/SPConstants.m]: Hungarian notation here as well ☝️

**Comment by jleandroperez** [comment]
Thank you Daniele!!!

**Pull request merged**